### PR TITLE
eth, eth/fetcher, eth/protocols/eth: retry dangling tx-fetch peers instead of excluding them forever

### PR DESCRIPTION
**Pull request by marcello33**
## Summary

`TxFetcher` permanently excluded a peer from all future transaction retrieval the moment a single `GetPooledTransactions` request to it went unanswered (the peer went "dangling"). There was no retry path, only a fresh peer connection or an incidental unrelated event could clear it. This is the root cause of POS-3689: some sentries with a full, healthy peer set sat with a near-empty local mempool for hours after a restart, because the one peer relationship whose request happened to go unanswered during the restart window was excluded from fetch-based retrieval permanently.

This PR adds real wire-level request/reply correlation so a dangling peer becomes eligible for retry after a bounded grace period instead of forever:

- `txRequest`/`txDelivery` now carry a `requestID`, threaded through the wire protocol's existing (previously unused) `RequestId` field on `GetPooledTransactionsPacket`/`PooledTransactionsPacket` — no wire-format change.
- A peer whose request times out (`txFetchTimeout`, 5s) is retried after `txFetchDanglingRetry` (7× timeout = 35s), rather than excluded forever.
- A late reply from the now-abandoned request is identified by requestID mismatch and discarded — but its content is still checked against announced metadata (protocol-violation enforcement) and its tracker state is still cleaned up, so a stale reply can't be used to dodge a legitimate drop.

## Executed tests

- New test functions in `eth/fetcher/tx_fetcher_test.go`, covering the dangling-retry lifecycle, requestID correlation, stale/unsolicited-delivery violation enforcement, and the same-peer hash-reuse edge case.
- `go test ./eth/... -race -count=1` — full pass, no races.
- `tests/fuzzers/txfetcher` clean.
- `diffguard -base develop` (mutation testing enabled) — stable at 86.0%.
- Live devnet reproduction (Kurtosis, 5 bor validators): injected a network fault on one validator's P2P port that pushed the tx-fetch round-trip past the 5s deadline while keeping the peer connection alive, and watched bor's own internal fetcher metrics (`eth_fetcher_transaction_slow_peers`, `_request_out/done/timeout` via `:7071/debug/metrics/prometheus`). Pre-fix (`develop`): metrics froze permanently once the fault cleared — 160+s with zero new fetch activity despite healthy peers and fresh transactions. Fixed build: kept retrying and issuing new requests after the fault cleared.

## Rollout notes

- Not consensus-affecting — pure P2P/mempool-propagation logic (`eth/fetcher`, `eth/protocols/eth`, plumbing in `eth/handler*.go`). No state, gas, genesis, or ABCI surface touched. Not hardfork-shaped.
- No wire-format change — `RequestId` already existed on the wire, just unused by the fetcher until now. No protocol version bump, no coordinated upgrade required.
- Backwards compatible; safe to roll out independently, no ordering constraint against peers on either code path.
- Operator-visible improvement only: a node should now recover mempool visibility for a stalled/half-alive peer automatically within ~20-25s instead of never.

### CODE REVIEW

File: eth/fetcher/tx_fetcher.go [hunk 1]
```diff
@@ -981,6 +1058,10 @@ func (f *TxFetcher) scheduleFetches(timer *mclock.Timer, timeout chan struct{},
 	}
 }
 
+func (f *TxFetcher) nextRequestID() uint64 {
+	return mrand.Uint64()
+}
```
[the remark this answers is not in the document]
Comment (marcello33) [reply eth/fetcher/tx_fetcher.go:1086]: Fixed, by adding a separate, independently-seeded requestIDs field just for request IDs (deterministic under a test seed, doesn't touch the shuffle order).

File: eth/fetcher/tx_fetcher.go [hunk 2]
```diff
@@ -136,7 +143,10 @@ type txDrop struct {
 //     state automata and there's do data leak.
```
[the remark this answers is not in the document]
Comment (marcello33) [reply eth/fetcher/tx_fetcher.go]: Fixed

File: eth/fetcher/tx_fetcher.go [hunk 3]
```diff
@@ -676,16 +705,27 @@ func (f *TxFetcher) loop() {
...
-				// Mark the requesting successful (independent of individual status)
+			// In case of a direct delivery that still resolves the currently
+			// tracked request, also reschedule anything missing from the
+			// original query. A stale request-ID delivery must not be
+			// treated as resolving a different, still-outstanding request.
+			if delivery.direct && !staleRequest {
+				// Mark the request successful (independent of individual status)
 				txRequestDoneMeter.Mark(int64(len(delivery.hashes)))
 
 				// Make sure something was pending, nuke it
```
[the remark this answers is not in the document]
Comment (marcello33) [reply eth/fetcher/tx_fetcher.go:736]: Confirmed against develop: this ordering is byte-for-byte identical there (not introduced by this PR). Leaving it alone to keep this PR scoped to the dangling-retry fix

File: eth/fetcher/tx_fetcher.go [hunk 4]
```diff
@@ -904,6 +941,36 @@ func (f *TxFetcher) rescheduleTimeout(timer *mclock.Timer, trigger chan struct{}
...
+	req := f.requests[peer]
+	if req == nil {
+		return true, false
+	}
+	if req.hashes != nil {
+		return false, false // genuinely in-flight, don't double-request
+	}
+	if time.Duration(now-req.time) < txFetchDanglingRetry {
+		return false, false // dangling, too recent to retry yet
+	}
```
[the remark this answers is not in the document]
Comment (marcello33) [reply eth/fetcher/tx_fetcher.go:989]: Changing that now would shift every dangling-retry test's timing. Fixed the doc comments on `txFetchDanglingRetry`/`evictStaleRequest` instead to accurately describe what the code does, so this doesn't read as a bug anymore.

File: eth/fetcher/tx_fetcher.go [hunk 5]
```diff
@@ -962,17 +1031,25 @@ func (f *TxFetcher) scheduleFetches(timer *mclock.Timer, timeout chan struct{},
...
-			f.requests[peer] = &txRequest{hashes: hashes, time: f.clock.Now()}
+			if wasDangling {
+				// Dangling placeholder actually superseded.
+				// Keep request-ID correlation by replacing it with
+				// the new request before the peer sees the new wire request.
+				txFetcherSlowPeers.Dec(1)
+				txFetcherSlowWait.Update(time.Duration(now - f.requests[peer].time).Nanoseconds())
+			}
+			requestID := f.nextRequestID()
+			f.requests[peer] = &txRequest{hashes: hashes, time: f.clock.Now(), requestID: requestID}
```
[the remark this answers is not in the document]
Comment (marcello33) [reply eth/fetcher/tx_fetcher.go:1062]: Fixed in latest commit.

**Review by vbhattaccmu — COMMENTED** [review]
The changes look good. One suggestion. We may set `txFetchDanglingRetry` above the 30 second [frameReadTimeout](https://github.com/0xPolygon/bor/blob/362beffa82fbaa98578a887dd36ac2a007a0b66c/p2p/server.go#L57). At 20 seconds the retry can replace `f.requests[peer]` while the original request is still live. A slow valid reply may then be checked against the wrong request ID. We can also add capped backoff or `h.jailPeer` strikes so an unresponsive peer is not retried forever.

File: eth/fetcher/tx_fetcher.go [hunk 6]
```diff
@@ -74,12 +74,28 @@ const (
...
+
+	// txFetchDanglingRetry is the minimum time since a request was originally
+	// sent (not since it went dangling at txFetchTimeout) before a peer whose
+	// request timed out without a reply becomes eligible for a fresh
+	// retrieval again. Kept above p2p's frameReadTimeout (30s): a peer whose
+	// underlying connection is actually dead gets dropped by then regardless,
+	// so a retry only ever races a reply from a peer that's merely slow, not
+	// disconnected. Request-ID correlation in the txDelivery direct-delivery
+	// path still handles a late reply from the superseded request either way.
+	txFetchDanglingRetry = 7 * txFetchTimeout
```
[the remark this answers is not in the document]
Comment (marcello33) [reply eth/fetcher/tx_fetcher.go:98]: Addressed in the PR description

**Review by vbhattaccmu — DISMISSED** [review]

File: packaging/deb/bor/DEBIAN/postrm [hunk 7]
```diff
@@ -2,17 +2,17 @@
...
-    *)
-        echo "postrm called with unknown argument '$1'" >&2
-        exit 1
-        ;;
+remove | purge | disappear)
+	if command -v systemctl >/dev/null 2>&1; then
+		systemctl daemon-reload >/dev/null 2>&1 || true
+	fi
+	;;
+upgrade | failed-upgrade | abort-install | abort-upgrade)
```
[the remark this answers is not in the document]
Comment (marcello33) [reply packaging/deb/bor/DEBIAN/postrm]: Resetting last commit introducing these changes

File: packaging/templates/package_scripts/postrm [hunk 8]
```diff
@@ -2,17 +2,17 @@
...
-    *)
-        echo "postrm called with unknown argument '$1'" >&2
-        exit 1
-        ;;
+remove | purge | disappear)
+	if command -v systemctl >/dev/null 2>&1; then
+		systemctl daemon-reload >/dev/null 2>&1 || true
+	fi
+	;;
+upgrade | failed-upgrade | abort-install | abort-upgrade)
```
[the remark this answers is not in the document]
Comment (marcello33) [reply packaging/templates/package_scripts/postrm]: Resetting last commit introducing these changes

**Review by vbhattaccmu — APPROVED** [review]

**Review by kamuikatsurgi — APPROVED** [review]
